Sources/StreamChat/Workers/MessageUpdater.swift

```diff
@@ -465,21 +465,18 @@ class MessageUpdater: Worker {
                 throw ClientError.MessageDoesNotExist(messageId: messageId)
             }
 
-            switch messageDTO.localMessageState {
-            case nil, .pendingSync, .syncingFailed, .deletingFailed:
-                try session.pin(message: messageDTO, pinning: pinning)
-                messageDTO.localMessageState = .pendingSync
-            case .pendingSend, .sendingFailed:
+            do {
                 try session.pin(message: messageDTO, pinning: pinning)
-                messageDTO.localMessageState = .pendingSend
-            case .sending, .syncing, .deleting:
-                throw ClientError.MessageEditing(
-                    messageId: messageId,
-                    reason: "message is in `\(messageDTO.localMessageState!)` state"
-                )
+            } catch {
+                log.warning("Failed to optimistically add the message pinning to the database: \(error)")
             }
-        }, completion: {
-            completion?($0)
+        }, completion: { error in
+            let endpoint: Endpoint<EmptyResponse> = .pinMessage(messageId: messageId, request: .init(set: .init(pinned: true)))
+            self.apiClient.request(endpoint: endpoint) { result in
+                guard let error = result.error else { return }
```

Comment:
Completion handler is skipped here which is unexpected.

Reply:
@laevandus, could you elaborate, please?

Reply:
How it looks currently:
```swift
func pinMessage(messageId: MessageId, pinning: MessagePinning, completion: ((Error?) -> Void)? = nil) {
        database.write({ session in
            guard let messageDTO = session.message(id: messageId) else {
                throw ClientError.MessageDoesNotExist(messageId: messageId)
            }

            try session.pin(message: messageDTO, pinning: pinning)
        }, completion: { [weak self] error in
            let endpoint: Endpoint<EmptyResponse> = .pinMessage(
                messageId: messageId,
                request: .init(set: .init(pinned: true))
            )

            self?.apiClient.request(endpoint: endpoint) { result in
                guard let error = result.error else { return }
                
                self?.database.write { session in
                    if let messageDTO = session.message(id: messageId) {
                        session.unpin(message: messageDTO)
                    }
                    completion?(error)
                }
            }
            completion?(error)
        })
    }
```

* If I call the pinMessage function it must trigger completion only once and never ignore calling it
* We trigger completion after write + after (apiRequest and another write) = 2 times
* The second write calls the completion from the write closure which is not correct, we should call it from the write's completion. If we call it from within the write closure, DB has not saved the change yet

Reply:
But we don't want to call the second database "write" on success 🤔

Please check out the latest commit, I copy-pasted the logic from `addReaction` and `deleteReaction` functions.

Reply:
I mixed up the previous snippet since on error we want to revert the state in the DB.

With the current implementation I am never receiving the error if the API request fails and the message stays as unpinned.
```swift
updater.pinMessage(messageId: "a", pinning: .xcz, completion: { error in
  if let error {
    // …
  } else {
    // … I expect that message is pinned here but it is not since API request failed
  }
}
```

Reply:
Updated, please check it out 🙂

Reply:
@laevandus, I can't understand how to bypass this swiftlint warning 🤔
`MessageUpdater.swift:484:1: warning: (hoistPatternLet) Reposition let or var bindings within pattern.`

This is about `case .failure(let apiError):`

Reply:
Will disable the rule `hoistPatternLet`